com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/LaunchRequestHandler.java

```diff
@@ -113,64 +122,153 @@
+                ((Connector.IntegerArgument) args.get("timeout")).setValue(ACCEPT_TIMEOUT);
+                String address = listenConnector.startListening(args);
+
+                String[] cmds = constructLaunchCommands(launchArguments, false, address);
+                RunInTerminalRequestArguments requestArgs = null;
+                if (launchArguments.console == CONSOLE.integratedTerminal) {
+                    requestArgs = RunInTerminalRequestArguments.createIntegratedTerminal(
+                            cmds,
+                            launchArguments.cwd,
+                            launchArguments.env,
+                            TERMINAL_TITLE);
+                } else {
+                    requestArgs = RunInTerminalRequestArguments.createExternalTerminal(
+                            cmds,
+                            launchArguments.cwd,
+                            launchArguments.env,
+                            TERMINAL_TITLE);
+                }
+                Request request = new Request(Command.RUNINTERMINAL.getName(),
+                        (JsonObject) JsonUtils.toJsonTree(requestArgs, RunInTerminalRequestArguments.class));
+
+                // Notes: In windows (reference to https://support.microsoft.com/en-us/help/830473/command-prompt-cmd--exe-command-line-string-limitation),
+                // when launching the program in cmd.exe, if the command line length exceed the threshold value (8191 characters),
+                // it will be automatically truncated so that launching in terminal failed. Especially, for maven project, the class path contains
+                // the local .m2 repository path, it may exceed the limit.
+                context.sendRequest(request, RUNINTERMINAL_TIMEOUT, (runResponse) -> {
+                    if (runResponse.success) {
+                        try {
+                            VirtualMachine vm = listenConnector.accept(args);
+                            context.setDebugSession(new DebugSession(vm));
+                            logger.info("Launching debuggee in terminal console succeeded.");
+                            resultFuture.complete(response);
+                        } catch (IOException | IllegalConnectorArgumentsException e) {
+                            logger.log(Level.SEVERE, String.format("Failed to launch debuggee in terminal. Reason: %s", e.toString()));
+                            resultFuture.complete(AdapterUtils.setErrorResponse(response, ErrorCode.LAUNCH_IN_TERMINAL_FAILURE,
+                                    String.format("Failed to launch debuggee in terminal. Reason: %s", e.toString())));
+                        }
+                    } else {
+                        logger.log(Level.SEVERE, String.format("Failed to launch debuggee in terminal. Reason: %s", runResponse.message));
+                        resultFuture.complete(AdapterUtils.setErrorResponse(response, ErrorCode.LAUNCH_IN_TERMINAL_FAILURE,
```

Comment:
I think we should call future.completeExceptionally(exception). Passing error using non-error interface is confusing.

Reply:
good suggestion.